File: languages/go/entity/encryption/encrypter.go
package encryption

import (
	"github.com/bloock/bloock-sdk-go/v2/entity/key"
	"github.com/bloock/bloock-sdk-go/v2/internal/bridge/proto"
)

// Encrypter represents an encryption configuration with various key types and access control.
type Encrypter struct {
	LocalKey           *key.LocalKey
	ManagedKey         *key.ManagedKey
	ManagedCertificate *key.ManagedCertificate
	LocalCertificate   *key.LocalCertificate
	AccessControl      *key.AccessControl
}

// NewEncrypterWithLocalKey creates a new Encrypter instance with a local encryption key.
func NewEncrypterWithLocalKey(key key.LocalKey) Encrypter {
	return Encrypter{
		LocalKey: &key,
	}
}

// NewEncrypterWithManagedKey creates a new Encrypter instance with a managed encryption key and access control.
func NewEncrypterWithManagedKey(key key.ManagedKey, accessControl *key.AccessControl) Encrypter {
	return Encrypter{
		ManagedKey:    &key,
		AccessControl: accessControl,
	}
}

// NewEncrypterWithLocalCertificate creates a new Encrypter instance with a local certificate for encryption.
func NewEncrypterWithLocalCertificate(key key.LocalCertificate) Encrypter {
	return Encrypter{
		LocalCertificate: &key,
	}
}

// NewEncrypterWithManagedCertificate creates a new Encrypter instance with a managed certificate for encryption and access control.
func NewEncrypterWithManagedCertificate(key key.ManagedCertificate, accessControl *key.AccessControl) Encrypter {
	return Encrypter{
		ManagedCertificate: &key,
		AccessControl:      accessControl,
	}
}

func (s Encrypter) ToProto() *proto.Encrypter {
	var localKey *proto.LocalKey
	if s.LocalKey != nil {
		localKey = s.LocalKey.ToProto()
	}

	var managedKey *proto.ManagedKey
	if s.ManagedKey != nil {
		managedKey = s.ManagedKey.ToProto()
	}

	var managedCertificate *proto.ManagedCertificate
	if s.ManagedCertificate != nil {
		managedCertificate = s.ManagedCertificate.ToProto()
	}

	var localCertificate *proto.LocalCertificate
	if s.LocalCertificate != nil {
		localCertificate = s.LocalCertificate.ToProto()
	}

	var accessControl *proto.AccessControl
	if s.AccessControl != nil {
		accessControl = s.AccessControl.ToProto()
	}

	return &proto.Encrypter{
		LocalKey:           localKey,
		ManagedKey:         managedKey,
		ManagedCertificate: managedCertificate,
		LocalCertificate:   localCertificate,
		AccessControl:      accessControl,
	}
}

File: languages/js/src/entity/encryption/encrypter.ts
import * as proto from "../../bridge/proto/encryption_entities";
import {
  LocalKey,
  ManagedKey,
  ManagedCertificate,
  LocalCertificate,
  AccessControl
} from "../key";

/**
 * Represents an encryption configuration with various key types and access control.
 */
export class Encrypter {
  localKey?: LocalKey;
  managedKey?: ManagedKey;
  managedCertificate?: ManagedCertificate;
  localCertificate?: LocalCertificate;
  accessControl?: AccessControl;

  /**
   * Creates a new Encrypter instance with a local encryption key.
   * @param key 
   * @param accessControl 
   */
  constructor(
    key: LocalKey | ManagedKey | ManagedCertificate | LocalCertificate,
    accessControl?: AccessControl
  ) {
    if (key instanceof LocalKey) {
      this.localKey = key;
    } else if (key instanceof ManagedKey) {
      this.managedKey = key;
    } else if (key instanceof ManagedCertificate) {
      this.managedCertificate = key;
    } else if (key instanceof LocalCertificate) {
      this.localCertificate = key;
    } else {
      throw new Error("invalid key provided");
    }

    if (accessControl) {
      this.accessControl = accessControl;
    }
  }

  public toProto(): proto.Encrypter {
    return proto.Encrypter.fromPartial({
      localKey: this.localKey?.toProto(),
      managedKey: this.managedKey?.toProto(),
      managedCertificate: this.managedCertificate?.toProto(),
      localCertificate: this.localCertificate?.toProto(),
      accessControl: this.accessControl?.toProto(),
    });
  }
}
